### ISSUE TITLE
conformer-ctc small not converge

### ISSUE THREAD

**Issue by BuaaAlban** [espnet/espnet#3760]
**Describe the bug**
Not a bug, ask for help and training techniques
conformer-ctc small can't  converge on Librispeech 960h

**Basic environments:**
 - OS information: Ubuntu 18.04.1 LTS
 - python version: [e.g. 3.8.5 (default, Sep 24 2020, 16:55:52)  [GCC 7.5.0]`]
 - espnet version: [e.g. espnet 0.9.6]
 - Git hash [e.g. c84da5743b7ef70c0c6212715859bdebdcf873b2]
   - Commit date [e.g. Tue Sep 1 09:32:54 2020 -0400]
 - pytorch version [e.g. pytorch 1.7.1]
 

**Environments from `torch.utils.collect_env`:**
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.1 LTS
GCC version: (GCC) 7.5.0
CMake version: version 3.10.2

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-PCIE
GPU 1: Tesla V100-PCIE
GPU 2: Tesla V100-PCIE
GPU 3: Tesla V100-PCIE

Nvidia driver version: 470.63.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] pytorch-wpe==0.0.1
[pip3] torch==1.7.1
[pip3] torch-complex==0.2.1
[conda] Could not collect

**Task information:**
 - Task: ASR
 - librispeech 960h
 - ESPnet2

**To Reproduce**
Actually I want to reproduce this model (https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_en_conformer_ctc_small) in espnet, but I can't get it converge.

I have tried several parameters, these are my config files. 
The  init one is the default conformer yaml in espnet and I change the ctc weight to 1.0 and the dimensions of encoder.
It can't converge and I tried to change the config it refering to (https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/conformer_ctc_bpe.yaml) 

[train_asr_conformer_ctc.txt](https://github.com/espnet/espnet/files/7495286/train_asr_conformer_ctc.txt)
[train_asr_conformer_ctc_v1.txt](https://github.com/espnet/espnet/files/7495283/train_asr_conformer_ctc_v1.txt)
[train_asr_conformer_ctc_v2.txt](https://github.com/espnet/espnet/files/7495284/train_asr_conformer_ctc_v2.txt)

I also find that the loss doesn't going down anymore at about epoch2 (maybe it's due to gradient vanishing?, the same happens for v0 and v2,  v1 diverge due to the large lr)
![loss_ctc](https://user-images.githubusercontent.com/25258181/140704211-79a0b392-d193-4828-8d47-4d9812e0dda4.png)

The training logs for v0 and v2 is here (there is warning for no valid stats in the log, does this have any impact on the results?)
[trainv0.log](https://github.com/espnet/espnet/files/7495365/trainv0.log)
[trainv2.log](https://github.com/espnet/espnet/files/7495368/trainv2.log)

I have run the whole process just for one time , the other exp are conducted from stage 10vim 

Do you have any suggestions ？ Thanks a lot if you could help

**Comment by sw005320** [comment]
We often observed such issues.

- Can you reduce the learning rate from the default (insetead of increasing it)?
- Can you also include the encod-decoder traininng? (make the ctc weight < 1.0) We found that the encod-decoder traininng well regularizes CTC training. We can just ignore them during the inference like the standard multi-task learning.

**Comment by BuaaAlban** [comment]
> We often observed such issues.
> 
> * Can you reduce the learning rate from the default (insetead of increasing it)?
> * Can you also include the encod-decoder traininng? (make the ctc weight < 1.0) We found that the encod-decoder traininng well regularizes CTC training. We can just ignore them during the inference like the standard multi-task learning.

Thanks for your suggestion, I have tried decreasing the lr from 0.001 to 0.0005 with ctc-weight=1.0 or 0.9(transformer decoder) and it didn't work, 
![image](https://user-images.githubusercontent.com/25258181/141071948-6e5f6699-0eed-462e-9a9d-27ae27f0aedc.png)
![image](https://user-images.githubusercontent.com/25258181/141071973-44b27995-ef3e-4184-81f2-d26454c8fd90.png)

I am doing an experience with lr=0.0001 now, hope it could converge.

**Comment by sw005320** [comment]
Hmm. I just want to make sure. Can you make it work with the original config (joint CTC/attention)?

**Comment by BuaaAlban** [comment]
> Hmm. I just want to make sure. Can you make it work with the original config (joint CTC/attention)?



> Hmm. I just want to make sure. Can you make it work with the original config (joint CTC/attention)?

Yes, it works well for the original ctc/attention model config in the repo on librispeech 960h.

But I want to try a smaller model with 144/176 dimension instead of 512, 4 attention heads instead of 8 attention heads, 16layers encoder instead of 12layers, ctc decoder instead of seq2seq (reproduce the nvidia conformer-ctc model )， and i wrote the config based on the original one.

I found it difficult to reproduce.

Or do you guys have tried to reproduce the conformer small model（seq2seq or Transducer） in the comformer paper? maybe that config can give insight to this conformer-ctc model.

**Comment by sw005320** [comment]
At least the CNN frontend part is different from the original one.
We use conv2d with 1/6 subsampling.

How about the BPE size?
This may make it difficult for CTC training.

Did you also adjust the kernel size?

We did not try small/medium/large conditions in the librispeech data.
We have such investigations in the aishell data.
@pengchengguo, could you also give some comments to @BuaaAlban?

**Comment by BuaaAlban** [comment]
> kernel
The sucessfully trained one (maybe a little different from the one in the repo) has something  different with the conformer-ctc.

The sucessfully trained  one using the default config have conv2d vs con2d6(mine)
BPE size are both 5000
kernel size are both 31 .
I noted that the accum_grad (1 vs 3  ) and batch_bins (140000000 vs 30000000) are different 
lr 0.0015 vs 0.001
init null vs xavier_uniform

I will try one config with only the layers/dim/ctc_weight/init difference

**Comment by BuaaAlban** [comment]
Besides I found the warning 
 "espnet/espnet2/train/reporter.py:91: UserWarning: No valid stats found
  warnings.warn("No valid stats found") "
is due to  ctc_weight=1.0

**Comment by pengchengguo** [comment]
Hi @BuaaAlban,

I am not familiar with Nvidia's Conformer model, but it is a little weird that reduces the attention dimension to 144/176 while keeping the FNN dimension as 2048. Do train w/ `Ir = 0.0001` and w/ `subsampling=conv2d` work?

Sorry, I didn't do experiments w/ different model sizes. For the Aishell experiments, I only tried relative positional type, the learning rate scheduler, and LayerNorm, etc.  

When having converge during training, I will try to use a smaller learning rate or kernel size. For the "No valid stats found" warning, I will check it ASAP.

**Comment by BuaaAlban** [comment]
> Hi @BuaaAlban,
> 
> I am not familiar with Nvidia's Conformer model, but it is a little weird that reduces the attention dimension to 144/176 while keeping the FNN dimension as 2048. Do train w/ `Ir = 0.0001` and w/ `subsampling=conv2d` work?
> 
> Sorry, I didn't do experiments w/ different model sizes. For the Aishell experiments, I only tried relative positional type, the learning rate scheduler, and LayerNorm, etc.
> 
> When having converge during training, I will try to use a smaller learning rate or kernel size. For the "No valid stats found" warning, I will check it ASAP.

Thanks a lot for your help and I found some feeling of tuning with your help.

The lr=0.0005 is still too big for this small model and lr =0.0001 seems working( though it's acc may not be that high, it converges at least.)

The model@epoch16 has about 80% acc. The followings are the figs for lr=0.0001 and ctcweight=0.7, the loss of the model with ctcweight=1.0 is similar, but I didn't put the fig because the acc fig is blank due to the **warning**, so the multi-task learning may influence the acc, but the lr decides whether it can converge.
![image](https://user-images.githubusercontent.com/25258181/141397535-fc5b041f-e509-4c63-9f4f-0d9de1f1b1d3.png)
![image](https://user-images.githubusercontent.com/25258181/141397562-3bc5126d-c1d0-4743-98b0-26d9e48829df.png)

Besides, you are right about FNN=2048, It should be expansion factor=4, which means 4*144. I will make an exp with it using a small lr like 0.0002 and conv2d instead of conv2d6

**Comment by BuaaAlban** [comment]
2021.11.15Update： The model has converged.  Thanks a lot for your help!
The following are the results.  
![image](https://user-images.githubusercontent.com/25258181/141749893-f92240c3-95a2-4da0-9269-5ebd2e7cb777.png)

@pengchengguo Do you have any update for the "warning" ? I will close this issue after your update

**Comment by pengchengguo** [comment]
Sorry, I haven't completely solved this problem, it works well on the Aishell 1 corpus. Does it only appear in the pure CTC training (ctc_weight=1) or also appear in joint ctc/attention training. I am preparing the librispeech data to reproduce this issue.

**Comment by BuaaAlban** [comment]
> Sorry, I haven't completely solved this problem, it works well on the Aishell 1 corpus. Does it only appear in the pure CTC training (ctc_weight=1) or also appear in joint ctc/attention training. I am preparing the librispeech data to reproduce this issue.

It only appears when ctc_weight=1 and is okay in  joint ctc/attention training.

**Comment by pengchengguo** [comment]
Hi @BuaaAlban,

For a pure CTC training, only `cer_ctc` will be computed. Both `acc_att` and `cer_attn` will be set to `None` (please refer to https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L171 for more detail). 

So you should modify the config file as
```
val_scheduler_criterion:
    - valid
    - cer_ctc
best_model_criterion:
    - valid
    - cer_ctc
    - min
```
for the pure CTC training.

**Comment by BuaaAlban** [comment]
@pengchengguo Thanks a lot for your guide, I will close the issue now.

**Comment by cuongld-vbee** [comment]
> 2021.11.15Update： The model has converged. Thanks a lot for your help! The following are the results. ![image](https://user-images.githubusercontent.com/25258181/141749893-f92240c3-95a2-4da0-9269-5ebd2e7cb777.png)
May you upload your config file for this result.

**Comment by BuaaAlban** [comment]
[train_asr_conformer_ctc_v5.txt](https://github.com/espnet/espnet/files/8037697/train_asr_conformer_ctc_v5.txt)
It should be this one, you can try it
